docs/auth_chain_difference_algorithm.md

```diff
@@ -0,0 +1,78 @@
+# Auth Chain Difference Algorithm
+
+The auth chain difference algorithm is used by V2 state resolution, where a
+naive implementation can be a significant source of CPU and DB usage.
+
+The auth chain difference of a set of state sets is the union minus the
+intersection of the sets of auth chains corresponding to the state sets, i.e an
+event is in the auth chain difference if it is reachable by walking the auth
+event graph from at least one of the state sets but not from *all* of the state
+sets.
+
+## Chain Cover Index
+
+Synapse computes auth chain differences by pre-computing a "chain cover" index
+for the auth chain in a room, allowing efficient reachability queries like "is
+event A in the auth chain of event B". This is done by assigning every event a
+*chain ID* and *sequence number* and having map of *links* such that A is
+reachable by B (i.e. `A` is in the auth chain of `B`) if and only if either:
+
+1. A and B have the same chain ID and `A`'s sequence number is less than `B`'s
+   sequence number; or
+2. there is a link `L` between `B`'s chain ID and `A`'s chain ID such that
+   `L.seq_no` <= `B.seq_no` and `A.seq_no` <= `L.seq_no`.
```

Comment:
We [talked about this a bit on #synapse-dev:matrix.org already](https://matrix.to/#/!XaqDhxuTIlvldquJaV:matrix.org/$FtS697tuUmnnDbtfufMsoP9qhP74i5xA7kFNjkkerKg?via=matrix.org&via=vector.modular.im&via=privacytools.io), but the `L.seq_no` references here are a bit confusing. It might also be clearer to define somewhere what makes up a link: 

> That's meant to be the start sequence number and end sequence number
>
>> there is a link `L` between `B`'s chain ID and `A`'s chain ID such that `L.start_seq_no` <= `B.seq_no` and `A.seq_no` <= `L.end_seq_no`.
>
> is probably what it should read I guess

And later:

> I guess I don't explain what links are terribly well. A link between `A` and `B` would be `(A.chain_id, A.seq_no)` -> `(B.chain_id, B.seq_no)`